src/enum_maps/base_enum.hpp

```diff
@@ -0,0 +1,52 @@
+/*
+	Copyright (C) 2008 - 2021
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include <string>
+#include <optional>
+
+template<class T>
+struct enum_base
+{
+	template<class Enum>
+	static std::string get_string(Enum key)
+	{
+		return T::values_[static_cast<int>(key)];
+	}
+
+	template<class Enum>
+	static std::optional<Enum> get_enum(const std::string& value)
+	{
+		for(unsigned int i = 0; i < T::values_.size(); i++)
+		{
+			if(value == T::values_[i])
+			{
```

Comment:
Please run this through clang-format (but be wary, it does bad changes to the macro definition).